File: backend/services/nearbyHospitals.js
/**
 * nearbyHospitals.js
 * Fetches real hospitals from OpenStreetMap Overpass API.
 * No API key required — completely free and open.
 */

const OVERPASS_URL = 'https://overpass-api.de/api/interpreter';

/**
 * Fetch real hospitals near a GPS coordinate.
 * @param {number} lat - Latitude
 * @param {number} lng - Longitude
 * @param {number} radiusKm - Search radius in km (default 10)
 * @returns {Promise<Array<{name, lat, lng, distance_km, address, phone, emergency, type}>>}
 */
async function fetchNearbyHospitals(lat, lng, radiusKm = 10) {
  const radiusM = radiusKm * 1000;

  const query = `
    [out:json][timeout:15];
    (
      node["amenity"="hospital"](around:${radiusM},${lat},${lng});
      way["amenity"="hospital"](around:${radiusM},${lat},${lng});
      node["amenity"="clinic"](around:${radiusM},${lat},${lng});
    );
    out center body;
  `;

  try {
    const response = await fetch(OVERPASS_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: `data=${encodeURIComponent(query)}`,
      signal: AbortSignal.timeout(15000),
    });

    if (!response.ok) throw new Error(`Overpass API error: ${response.status}`);
    const data = await response.json();

    const hospitals = (data.elements || [])
      .filter(el => el.tags && el.tags.name)
      .map(el => {
        const elLat = el.lat || el.center?.lat;
        const elLng = el.lon || el.center?.lon;
        if (!elLat || !elLng) return null;

        const dist = haversineDistance(lat, lng, elLat, elLng);

        return {
          id: `osm_${el.id}`,
          name: el.tags.name,
          lat: elLat,
          lng: elLng,
          distance_km: Math.round(dist * 10) / 10,
          address: el.tags['addr:full'] || el.tags['addr:street'] || '',
          phone: el.tags.phone || el.tags['contact:phone'] || '',
          emergency: el.tags.emergency === 'yes' || el.tags.amenity === 'hospital' ? 'yes' : 'no',
          type: el.tags.amenity === 'hospital' ? 'Hospital' : 'Clinic',
          dept: el.tags.healthcare || 'General',
        };
      })
      .filter(Boolean)
      .sort((a, b) => a.distance_km - b.distance_km)
      .slice(0, 15); // Top 15 nearest

    return hospitals;
  } catch (error) {
    console.error('Overpass API error:', error.message);
    // Fallback: return empty array, frontend handles gracefully
    return [];
  }
}

/**
 * Haversine formula — distance between two GPS points in km.
 */
function haversineDistance(lat1, lon1, lat2, lon2) {
  const R = 6371;
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) *
    Math.sin(dLon / 2) * Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

function toRad(deg) { return deg * (Math.PI / 180); }

module.exports = { fetchNearbyHospitals, haversineDistance };

File: backend/services/nearby_hospitals.py
"""
nearby_hospitals.py — Fetch real hospitals from OpenStreetMap Overpass API.
Mirrors services/nearbyHospitals.js.
"""
import math
from typing import Optional
import httpx

_OVERPASS_URL = "https://overpass-api.de/api/interpreter"


def _haversine(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    R = 6371
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = math.sin(dlat / 2) ** 2 + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(dlon / 2) ** 2
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


async def fetch_nearby_hospitals(lat: float, lng: float, radius_km: float = 10) -> list[dict]:
    """
    Query OpenStreetMap Overpass API for hospitals/clinics near (lat, lng).
    Returns a list of facility dicts sorted by distance.
    """
    radius_m = radius_km * 1000
    query = f"""
[out:json][timeout:25];
(
  node["amenity"="hospital"](around:{radius_m},{lat},{lng});
  node["amenity"="clinic"](around:{radius_m},{lat},{lng});
  node["amenity"="health_post"](around:{radius_m},{lat},{lng});
  node["amenity"="doctors"](around:{radius_m},{lat},{lng});
  node["healthcare"="centre"](around:{radius_m},{lat},{lng});
  node["healthcare"="hospital"](around:{radius_m},{lat},{lng});
  node["healthcare"="clinic"](around:{radius_m},{lat},{lng});
  way["amenity"="hospital"](around:{radius_m},{lat},{lng});
  way["healthcare"="hospital"](around:{radius_m},{lat},{lng});
);
out center;
""".strip()

    try:
        async with httpx.AsyncClient(timeout=22) as client:
            resp = await client.post(
                _OVERPASS_URL,
                content=f"data={httpx.QueryParams({'': query}).value.lstrip('=')}",
                headers={"Content-Type": "application/x-www-form-urlencoded"},
            )
            resp.raise_for_status()
            data = resp.json()
    except Exception as e:
        print(f"[Overpass] Error: {e}")
        return []

    facilities = []
    for el in data.get("elements", []):
        tags = el.get("tags", {})
        if not tags.get("name"):
            continue
        f_lat = el.get("lat") or (el.get("center") or {}).get("lat")
        f_lng = el.get("lon") or (el.get("center") or {}).get("lon")
        if not f_lat or not f_lng:
            continue
        dist = round(_haversine(lat, lng, f_lat, f_lng), 2)
        facilities.append({
            "id": str(el.get("id")),
            "name": tags["name"],
            "lat": f_lat,
            "lng": f_lng,
            "distance_km": dist,
            "address": ", ".join(filter(None, [
                tags.get("addr:full"),
                tags.get("addr:street"),
                tags.get("addr:city"),
            ])) or "",
            "phone": tags.get("phone") or tags.get("contact:phone") or "",
            "emergency": tags.get("emergency", "unknown"),
            "type": tags.get("amenity") or tags.get("healthcare") or "health_facility",
        })

    facilities.sort(key=lambda x: x["distance_km"])
    return facilities[:30]
